pkg/tenant/tenants.go

```diff
@@ -0,0 +1,121 @@
+package tenant
+
+import (
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/oklog/ulid/v2"
+	"github.com/rotationalio/ensign/pkg/tenant/api/v1"
+	"github.com/rotationalio/ensign/pkg/tenant/db"
+	"github.com/rs/zerolog/log"
+)
+
+// CreateTenant creates a new tenant in the database
+func (s *Server) CreateTenant(c *gin.Context) {
+	var (
+		err    error
+		tenant *api.Tenant
+	)
+
+	if err = c.BindJSON(&tenant); err != nil {
+		log.Warn().Err(err).Msg("could not parse tenant create request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse request"))
+		return
+	}
+
+	if tenant.ID == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(api.ErrTenantIDRequired))
+	}
+
+	if tenant.Name == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("tenant name is required"))
+	}
+
+	if tenant.EnvironmentType == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("environment type is required"))
+	}
+
+	// Add Tenant to the database
+	if err = db.CreateTenant(c.Request.Context(), &db.Tenant{ID: ulid.Make()}); err != nil {
+		log.Error().Err(err).Msg("could not save tenant")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add tenant"))
+		return
+	}
+
+	c.JSON(http.StatusCreated, tenant)
+}
+
+func (s *Server) TenantList(c *gin.Context) {
+
+	c.JSON(http.StatusNotImplemented, "not implemented yet")
+}
+
+func (s *Server) TenantCreate(c *gin.Context) {
+	var (
+		err    error
+		tenant *api.Tenant
+	)
+
+	// TODO: Add authentication and authorization middleware
+	if err = c.BindJSON(&tenant); err != nil {
+		log.Warn().Err(err).Msg("could not parse tenant create request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not bind request"))
+	}
+
+	if tenant.ID == "" {
+		log.Error().Err(err).Msg("tenant does not have an ID")
+	}
+
+	if tenant.Name == "" {
+		log.Warn().Err(err).Msg("tenant does not have an name")
+	}
+
+	if tenant.EnvironmentType == "" {
+		log.Error().Err(err).Msg("tenant does not have an environment type")
+	}
+
+	// TODO: Add tenant to the database
+	if err = db.CreateTenant(c.Request.Context(), &db.Tenant{ID: ulid.Make()}); err != nil {
+		log.Error().Err(err).Msg("could not save tenant")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add tenant"))
+	}
+
+	c.JSON(http.StatusCreated, tenant)
+}
+
+func (s *Server) TenantDetail(c *gin.Context) {
+	c.JSON(http.StatusNotImplemented, "not implemented yet")
+}
+
+func (s *Server) TenantUpdate(c *gin.Context) {
+	c.JSON(http.StatusNotImplemented, "not implemented yet")
+}
+
+func (s *Server) TenantDelete(c *gin.Context) {
+	var (
+		err error
+		out *api.Reply
+	)
+
+	tenantID := c.Param("tenantID")
+
+	if tenantID == "" {
+		log.Error().Err(err).Msg(db.ErrMissingID.Error())
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not retrieve tenant"))
+		return
+	}
```

Comment:
You're pretty much guaranteed to get some string as the `tenantID` otherwise this handler would not be called. However, if the tenantID is not a ULID, we should return a 404 since the parameter is part of the url. E.g. if the request is to `DELETE /v1/tenants/foo` -- `foo` is a 404 since it's not a ULID. 

```suggestion
	tenantID := c.Param("tenantID")
	var tenantID ulid.ULID
	if tenantID, err = ulid.Parse(c.Param("tenantID")); err != nil {
		log.Debug().Err(err).Msg("could not parse tenant ulid")
		c.JSON(http.StatusNotFound, api.ErrorResponse(ErrTenantNotFound))
		return
	}
```

Reply:
While dealing with an error (the one related to getting the mock) I decided to test a delete method to see if I'd get the same issue. I thought I'd deleted this in my last commit. Oh well. I'll save this for when I review the detail and delete changes before submitting a PR.